### DISCUSSION TITLE
Normalize consecutive spaces?

### DISCUSSION THREAD

**Discussion by bpj** [jgm/pandoc#10399] [Q&A]
Didn't there use to be an option/extension to normalize consecutive spaces and the like in output?

My use case is that I have a filter which removes certain spans, and I end up with double spaces in the output where there are spaces before and after those spans.

I'm just wondering as I seem to remember this. It is no big deal as the obvious workaround is to put the leading space inside the span, which I probably should do anyway if there is a terminal punctuation after the span.

**Comment by jgm** [comment]
When you remove the spans, are you replacing them with an empty span or with nothing `{}`?
In the latter case I would not expect you'd get doubled spaces in the output.
```
% pandoc -f native -t html
[Str "hi",Space,Space,Str "hi"]
hi hi
```

**Reply by bpj** [reply]
I replace them with nothing.

My code is this (comments added):

``````lua
local c = elem.attributes["if"]
if c then
  if include[c] then
    if wrap[c] then
      return nil -- unchanged
    end
    -- else unwrap
    return elem.content
  end
  -- elseif not include[c]
  return { }
end
``````
Strange!

**Reply by jgm** [reply]
What does `native` output look like after your filter? Are you seeing `Space, Space`?  If so that should be normalized in most formats automatically. (Maybe not all.)

**Reply by bpj** [reply]
Yes there are two spaces in native output, but I have narrowed it down further: the context where both spaces are preserved is inside a heading with the markdown writer.

**Reply by jgm** [reply]
Give the native representation of the heading so I can test.
It should still be collapsed in HTML output.

**Reply by bpj** [reply]
``````code
# foo [bar]{if=a} baz

[ Header
    1
    ( "foo-bar-baz" , [] , [] )
    [ Str "foo" , Space , Space , Str "baz" ]
]

# foo  baz {#foo-bar-baz}

<h1 id="foo-bar-baz">foo baz</h1>
``````

As I said it happens in markdown output, but not in html output.

**Reply by jgm** [reply]
Sorry, I missed the part about markdown output.

**Reply by jgm** [reply]
We use a `nowrap` combinator for headers, to make sure they're all on a single line, and that may be affecting this...will need to remind myself how that works in doclayout.

**Reply by jgm** [reply]
In any case, it should be possible to write a Lua filter that collapses adjacent Space elements.
use `Inlines` and iterate over the list.

**Reply by bpj** [reply]
I will probably do that in a second pass. I have anyway found that it is better to put spaces which go before the content of conditional spans inside the span, or I may end up with spaces before punctuation.

**Reply by jgm** [reply]
It is `nowrap` that is doing this.  The implementation could be improved.

**Reply by bpj** [reply]
Do you want me to close this and/or open an issue?

**Reply by jgm** [reply]
It's a pretty low priority thing, but if you want to open an issue, you could open it on jgm/doclayout.
The issue would be that `nowrap` simply converts breaking spaces (which normally collapse) into literal spaces, so that two adjacent ones will produce two spaces.
